federation-gateway-bundle/feg-hello-operator/tox.ini

```
@@ -0,0 +1,89 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+[tox]
+skipsdist=True
+skip_missing_interpreters = True
+envlist = lint, static, unit
+
+[vars]
+src_path = {toxinidir}/src/
+unit_test_path = {toxinidir}/tests/unit/
+integration_test_path = {toxinidir}/tests/integration/
+all_path = {[vars]src_path} {[vars]unit_test_path} {[vars]integration_test_path}
+
+[testenv]
+deps = 
+    pytest
+    pytest-operator
+setenv =
+    PYTHONPATH = {toxinidir}:{toxinidir}/lib:{[vars]src_path}
+    PYTHONBREAKPOINT=ipdb.set_trace
+passenv =
+    HTTP_PROXY
+    HTTPS_PROXY
+    NO_PROXY
+    PYTHONPATH
+    HOME
+    CHARM_BUILD_DIR
+    MODEL_SETTINGS
+
+[testenv:fmt]
+description = Apply coding style standards to code
+deps =
+    black
+    isort
+commands =
+    isort {[vars]all_path}
+    black {[vars]all_path}
+
+[testenv:lint]
+description = Check code against coding style standards
+deps =
+    black
+    flake8
+    flake8-docstrings
+    flake8-copyright
+    flake8-builtins
+    pyproject-flake8
+    pep8-naming
+    isort
+commands =
+    pflake8 {[vars]all_path}
+    isort --check-only --diff {[vars]all_path}
+    black --check --diff {[vars]all_path}
+
+[testenv:static]
+description = Run static analysis checks
+deps =
+    -r{toxinidir}/requirements.txt
+    mypy
+    types-PyYAML
+    pytest
+    pytest-operator
+    juju
+    types-setuptools
+    types-toml
+commands =
+    mypy {[vars]all_path} {posargs}
+
+[testenv:unit]
+description = Run unit tests
+deps =
+    pytest
+    coverage[toml]
+    -r{toxinidir}/requirements.txt
+commands =
+    coverage run --source={[vars]src_path} -m pytest --ignore {[vars]integration_test_path} -v --tb native -s {posargs}
+    coverage report
+
+[testenv:integration]
+description = Run integration tests
+deps =
+    git+https://github.com/juju/python-libjuju.git
+    pytest
+    git+https://github.com/charmed-kubernetes/pytest-operator.git
+    -r{toxinidir}/requirements.txt
+commands =
+    sudo apt-get --no-install-recommends -y -oDpkg::Use-Pty=0 --allow-downgrades install python3.8-venv=3.8.10-0ubuntu1~20.04.4
```

This version does not exist in the repos anymore. It is better to remove the version.
```suggestion
    sudo apt-get --no-install-recommends -y -oDpkg::Use-Pty=0 --allow-downgrades install python3.8-venv
```

---

Done